smart-contract-verifier/smart-contract-verifier/src/batch_verifier/batch_contract_verifier.rs

```diff
@@ -0,0 +1,245 @@
+    Failure(Vec<VerificationError>),
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct BatchSuccess {
+    pub compiler: String,
+    pub compiler_version: String,
+    pub language: String,
+    pub compiler_settings: serde_json::Value,
+    pub creation_code: Vec<u8>,
+    pub runtime_code: Vec<u8>,
+    pub file_name: String,
+    pub contract_name: String,
+    pub sources: BTreeMap<String, String>,
+    pub compilation_artifacts: serde_json::Value,
+    pub creation_code_artifacts: serde_json::Value,
+    pub runtime_code_artifacts: serde_json::Value,
+    pub creation_match: Option<Match>,
+    pub runtime_match: Option<Match>,
+}
+
+pub async fn verify_solidity(
+    compilers: &Compilers<SolidityCompiler>,
+    compiler_version: Version,
+    contracts: Vec<Contract>,
+    compiler_input: &foundry_compilers::CompilerInput,
+) -> Result<Vec<VerificationResult>, BatchError> {
+    let (raw_compiler_output, _) = compilers
+        .compile(&compiler_version, compiler_input, None)
+        .await?;
+
+    let (modified_raw_compiler_output, _) = {
+        let compiler_input = compiler_input.clone().modify();
+        compilers
+            .compile(&compiler_version, &compiler_input, None)
+            .await?
+    };
+
+    let compilation_result = compilation::parse_solidity_contracts(
+        compiler_version,
+        compiler_input,
+        raw_compiler_output,
+        modified_raw_compiler_output,
+    )
+    .map_err(|err| {
+        tracing::error!("parsing compiled contracts failed: {err:#}");
+        BatchError::Internal(err)
+    })?;
+
+    let mut results = vec![];
+    for contract in contracts {
+        results.push(verify_contract(contract, &compilation_result)?);
+    }
+
+    Ok(results)
+}
+
+fn verify_contract(
+    contract: Contract,
+    compilation_result: &CompilationResult,
+) -> Result<VerificationResult, BatchError> {
+    let mut successes: Vec<BatchSuccess> = Vec::new();
+    let mut failures: Vec<VerificationError> = Vec::new();
+    for parsed_contract in &compilation_result.parsed_contracts {
+        let convert_error = |mut kind, context: &'static str| {
+            if let VerificationErrorKind::InternalError(err) = kind {
+                kind = VerificationErrorKind::InternalError(err.context(context))
+            }
+            VerificationError::new(
+                parsed_contract.file_name.clone(),
+                parsed_contract.contract_name.clone(),
+                kind,
+            )
+        };
+
+        let (does_creation_match, creation_values, creation_transformations) = match &contract
+            .creation_code
+        {
+            Some(contract_code) => {
+                match transformations::process_creation_code(
+                    contract_code,
+                    parsed_contract.creation_code.to_vec(),
+                    &parsed_contract.compilation_artifacts,
+                    serde_json::to_value(parsed_contract.creation_code_artifacts.clone()).unwrap(),
```

Comment:
I suggest to replace this logic of converting to json and parsing json with enum :)